packages/gatsby/lib/query-runner/page-query-runner.js

```diff
@@ -0,0 +1,120 @@
+/***
+ * Jobs of this module
+ * - Ensure on bootstrap that all invalid page queries are run and report
+ *   when this is done
+ * - Watch for when a page's query is invalidated and re-run it.
+ ***/
+
+const _ = require("lodash")
+const Promise = require("bluebird")
+
+const { store } = require("../redux")
+const queryRunner = require("./query-runner")
+const checkpointsPromise = require("../utils/checkpoints-promise")
+
+let queuedDirtyActions = []
+// Don't start running queries until we've reached the BOOTSTRAP_STAGE of
+// QUERY_RUNNER.
+let active = false
+const callbacks = []
+module.exports = cb => {
+  callbacks.push(cb)
+}
+
+checkpointsPromise({
+  events: [`COMPONENT_QUERIES_EXTRACTION_FINISHED`],
+}).then(() => {
+  active = true
+  const state = store.getState()
+
+  // Run queued dirty nodes now that we're active.
+  queuedDirtyActions = _.uniq(queuedDirtyActions, a => a.payload.id)
+  findAndRunQueriesForDirtyPaths(queuedDirtyActions).then(() => {
+    // Find paths without data dependencies and run them (just in case?)
+    const paths = findPathsWithoutDataDependencies()
+    // Run these pages
+    paths.forEach(path => {
+      const page = state.pages.find(p => p.path === path)
+      const component = state.pageComponents[page.component]
+      queryRunner(page, component)
+    })
+
+    // Tell everyone who cares that we're done.
+    callbacks.forEach(cb => cb())
```

Comment:
I think we can have a look into [ionic-dev-tools ](https://github.com/driftyco/ionic-app-scripts), they've got many of our problematics (live-reload, watch, build, notifications…) and their code is very clean ;)